src/trace/entry.zig

```diff
@@ -1,245 +1,87 @@
 const std = @import("std");
-const logfmt = @import("logfmt.zig");
-const Level = @import("level.zig").Level;
-const Channel = @import("../sync/channel.zig").Channel;
+const sig = @import("../sig.zig");
+const trace = @import("lib.zig");
+
+const Level = trace.level.Level;
+const ScopedLogger = trace.log.ScopedLogger;
 const AtomicBool = std.atomic.Value(bool);
 
-pub const Entry = union(enum) {
-    channel_print: ChannelPrintEntry,
-    direct_print: DirectPrintEntry,
-    noop,
+pub fn NewEntry(comptime scope: ?[]const u8) type {
+    return Entry(struct {}, scope);
+}
 
-    const Self = @This();
+pub fn Entry(comptime Fields: type, comptime scope: ?[]const u8) type {
+    return struct {
+        logger: ScopedLogger(scope),
+        level: Level,
+        fields: Fields,
 
-    pub fn init(allocator: std.mem.Allocator, channel: *Channel(logfmt.LogMsg), log_level: Level) Self {
-        return .{ .channel_print = ChannelPrintEntry.init(allocator, channel, log_level) };
-    }
+        const Self = @This();
 
-    pub fn deinit(self: Self) void {
-        switch (self) {
-            .noop => {},
-            .channel_print => |impl| impl.deinit(),
-            .direct_print => |impl| impl.deinit(),
+        pub inline fn init(logger: ScopedLogger(scope), level: Level) Self {
+            return .{ .logger = logger, .level = level, .fields = .{} };
         }
-    }
 
-    pub fn field(self: Self, name: []const u8, value: anytype) Self {
-        switch (self) {
-            .noop => {
-                return self;
-            },
-            .channel_print => |entry| {
-                var log_entry = entry;
-                _ = log_entry.field(name, value);
-                return Entry{ .channel_print = log_entry };
-            },
-            .direct_print => |entry| {
-                var log_entry = entry;
-                _ = log_entry.field(name, value);
-                return Entry{ .direct_print = log_entry };
-            },
+        /// Add a field to the log message.
+        pub inline fn field(
+            self: Self,
+            comptime name: [:0]const u8,
+            value: anytype,
+        ) Entry(FieldsPlus(name, @TypeOf(value)), scope) {
```

Comment:
Interesting function types can be used this way.

Reply:
Is it clear what's going on here?

Reply:
It is. But that is a bit of a biased answer given the familiarity with this part of the code the past couple of weeks.

Reply:
I think it's be fair to say that how this code works may not be obvious to someone who doesn't understand that, in zig, types can be computed through arbitrary logic just as any other values.
But I think understanding that facet of comptime is a fair prerequisite to reading sig, as long as we don't get much more complicated than this.